Sources/ApolloCodegenLib/Frontend/JavaScriptBridge.swift

```diff
@@ -0,0 +1,345 @@
+import JavaScriptCore
+
+// JavaScriptCore APIs haven't been annotated for nullability, but most of its methods will never return `null`
+// and can be safely force unwrapped. (Even when an exception is thrown they would still return
+// a `JSValue` representing a JavaScript `undefined` value.)
+
+/// An errror thrown during JavaScript execution.
+/// See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error
+public class JavaScriptError: JavaScriptObject, Error {
+  lazy var name: String? = self["name"]
+  
+  lazy var message: String? = self["message"]
+  
+  lazy var stack: String? = self["stack"]
+}
+
+extension JavaScriptError: CustomStringConvertible {
+  public var description: String {
+    return jsValue.toString()
+  }
+}
+
+/// A type that references an underlying JavaScript object.
+public class JavaScriptObject: JavaScriptValueDecodable {
+  let jsValue: JSValue
+  unowned let bridge: JavaScriptBridge
+  
+  static func fromJSValue(_ jsValue: JSValue, bridge: JavaScriptBridge) -> Self {
+    return bridge.wrap(jsValue)
+  }
+    
+  required init(_ jsValue: JSValue, bridge: JavaScriptBridge) {
+    precondition(jsValue.isObject)
+        
+    self.jsValue = jsValue
+    self.bridge = bridge
+  }
+  
+  subscript(property: Any) -> JSValue {
+    return jsValue[property]
+  }
+  
+  subscript<Decodable: JavaScriptValueDecodable>(property: Any) -> Decodable {
```

Comment:
this seems like it should throw - what happens if the property doesn't exist?

Reply:
Subscripts can't be throwable in Swift. And this is similar to the other cases. You have to trust that the correspondence between JavaScript and Swift has been set up correctly. If something goes wrong, you'll still get a precondition failure, but it's not an error you can do anything about.

Reply:
is it worth making this an explicit `fatalError` with a more readable error message?

Reply:
The preconditions already contain readable error messages.